drivers/lcd/lcd_dev.c

```diff
@@ -123,20 +123,56 @@ static int lcddev_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
     {
     case LCDDEVIO_GETRUN:
       {
-        FAR struct lcddev_run_s *lcd_putrun =
+        FAR struct lcddev_run_s *lcd_run =
             (FAR struct lcddev_run_s *)arg;
 
-        ret = priv->planeinfo.getrun(lcd_putrun->row, lcd_putrun->col,
-                                     lcd_putrun->data, lcd_putrun->npixels);
+        ret = priv->planeinfo.getrun(lcd_run->row, lcd_run->col,
+                                     lcd_run->data, lcd_run->npixels);
       }
       break;
     case LCDDEVIO_PUTRUN:
       {
-        const FAR struct lcddev_run_s *lcd_putrun =
+        const FAR struct lcddev_run_s *lcd_run =
             (const FAR struct lcddev_run_s *)arg;
 
-        ret = priv->planeinfo.putrun(lcd_putrun->row, lcd_putrun->col,
-                                     lcd_putrun->data, lcd_putrun->npixels);
+        ret = priv->planeinfo.putrun(lcd_run->row, lcd_run->col,
+                                     lcd_run->data, lcd_run->npixels);
+      }
+      break;
+    case LCDDEVIO_PUTAREA:
+      {
+        const FAR struct lcddev_area_s *lcd_area =
+            (const FAR struct lcddev_area_s *)arg;
+
+        if (priv->planeinfo.putarea)
+          {
+            ret = priv->planeinfo.putarea(lcd_area->row_start,
```

Comment:
need optimize lcd_framebuffer.c too, maybe we can write a common function lcd_putarea(...) shared by lcd_dev.c and lcd_framebuffer.c

Reply:
It depend on how the GUI library implement. framebuffer may fast than putrun/putarea for the advantage animation or a alpha blending. Yes, framebuffer consume more memory, but some project more care about the visual experience than resource.

Reply:
But the LCD framebuffer driver simply emulates a framebuffer over LCD operations. In fact, with the putarea() interface, you can achieve the same behavior as the LCD framebuffer has now, which blits the entire buffer using putrun() for each call. So, you can get the same results by just doing putarea() over the whole display whenever you want.

Reply:
Here is why the speed may fast than putarea:

1. Some GUI simply draw widgets from bottom to top which mean the same area will draw multiple times. Internal RAM is much fast than SPI LCD: Write twice RAM plus one SPI transfer normally fast than two SPI transfer
2. Advance UI(e.g. alpha blending) normally need read back the image data. LCD either don't support getarea or the speed is much slower than RAM.

Reply:
I understand, but all of these are reasons to have a buffer instead of directly using LCD operations every time something needs to be drawn, not reasons for the LCD framebuffer itself. Of course this driver offers a buffer, but any application can simply create a buffer, draw there and then simply use the LCD driver with one putarea() for any/all affected portions of the buffer (in fact, this choice is not available in the LCD framebuffer, it always transfers the whole image).

The only reason for this LCD framebuffer driver would be for an application written for a mmap'ed framebuffer interface which you don't want to modify. But do we have such concern?

Reply:
> I understand, but all of these are reasons to have a buffer instead of directly using LCD operations every time something needs to be drawn, not reasons for the LCD framebuffer itself. Of course this driver offers a buffer, but any application can simply create a buffer,

If the framebuffer is required in some case, why not directly use lcd_framebuffer.c directly? For example, now LVGL support framebuffer interface(/dev/fbx), and most likely will support LCD interface(/dev/lcdx). Do you plan to support the third combination(userspace framebuffer + /dev/lcdx)? or reuse the lcd_framebuffer.c + /dev/fbx? The later design is much better, do you think so?

> draw there and then simply use the LCD driver with one putarea() for any/all affected portions of the buffer (in fact, this choice is not available in the LCD framebuffer, it always transfers the whole image).
> 

No LCD framebuffer provide the interface to let GUI library control which portion of framebuffer need update:
https://github.com/apache/incubator-nuttx/blob/master/drivers/lcd/lcd_framebuffer.c#L176

> The only reason for this LCD framebuffer driver would be for an application written for a mmap'ed framebuffer interface which you don't want to modify. But do we have such concern?

I really care about is how fast LVGL with the advanced UI(alpha blending and smooth animation) can run on LCD panel. In this case， the framebuffer plus lcd is more suitable than pure lcd interface. lcd_framebuffer.c can work immediately with LVGL porting.

Reply:
So, what I expect is that:

1. For resource constraint system, user enable CONFIG_LCD_DEV and LVGL open /dev/lcdx
2. For advanced UI, user enable CONFIG_LCD_FRAMEBUFFER and LVGL open /dev/fbx

Reply:
> If the framebuffer is required in some case, why not directly use lcd_framebuffer.c directly? For example, now LVGL support framebuffer interface(/dev/fbx), and most likely will support LCD interface(/dev/lcdx). 

I'm not sure which combination you are referring to and what do you mean by "LVGL support". LVGL does not care about the interface, it simply draws to an internal buffer and you supply a callback which applies the changed portion to the hardware. If you refer to the lvgldemo, this is the part that we write ourselves (not part of LVGL). Right now indeed it supports framebuffer and it is trivial to add support for LCD. I'm working extensively with LVGL for some time and I've been using the LCD interface in this same way, precisely because the LCD framebuffer device requires so much memory (consider this: 176x176 pixels: 30KB of RAM, that is quite a lot to have duplicated).

> Do you plan to support the third combination(userspace framebuffer + /dev/lcdx)? or reuse the lcd_framebuffer.c + /dev/fbx? The later design is much better, do you think so?

Yes, I can easily add this option, consider this code:

```
static void flush_cb(lv_disp_drv_t* disp_drv, const lv_area_t * area, lv_color_t * color_p)
{
#ifdef CONFIG_VIDEO_FB
  lv_color_t* fb = (lv_color_t*)((bc::Display*)disp_drv->user_data)->framebuffer_ptr;
  for (int row = area->y1; row <= area->y2; row++)
    {
      for (int col = area->x1; col <= area->x2; col++, color_p++)
        {
          fb[row * CONFIG_LVGL_HOR_RES_MAX + col] = (lv_color_to1(*color_p) == 1 ? LV_COLOR_WHITE : LV_COLOR_BLACK);
        }
    }
#else
  struct lcddev_putrun_s putrun;
  putrun.col = area->x1;
  putrun.npixels = (area->x2 - area->x1) + 1;
  putrun.data = (uint8_t*)color_p;

  bc::Display* display = (bc::Display*)disp_drv->user_data;

  for (int row = area->y1; row <= area->y2; row++)
    {
      putrun.row = row;
      ioctl(display->fd, LCDDEVIO_PUTRUN, (unsigned long)&putrun);
      putrun.data += putrun.npixels >> 3;
    }
#endif

  lv_disp_flush_ready(disp_drv);
}
```

That is the only thing that needs touching as far as LVGL goes. The other difference is in the opening of the LCD driver instead of the FB (and no mmap is needed).

Note that the above code does not take advantage of putrun(), with putarea() it now becomes really efficient. LVGL only invokes this for modified areas, so not double drawing. Also, note that when using LVGL you always supply a buffer where to do drawing operations (it can be as big as your screen or less, and LVGL accomodates with some extra drawing).

In this scenario, LCD framebuffer is simply a kludge, it adds an extra buffer which is not necessary. I've tried this and it is *much* slower (and remind that there was a recent issue opened by @w8jcik regarding memory use).

> No LCD framebuffer provide the interface to let GUI library control which portion of framebuffer need update:
> https://github.com/apache/incubator-nuttx/blob/master/drivers/lcd/lcd_framebuffer.c#L176

I haven't seen that interface, it didn't existed before. So it boils down to extra memory use (besides the fact of putting one driver on top of the other).

> > The only reason for this LCD framebuffer driver would be for an application written for a mmap'ed framebuffer interface which you don't want to modify. But do we have such concern?
> 
>I really care about is how fast LVGL with the advanced UI(alpha blending and smooth animation) can run on LCD panel. In this case, framebuffer plus lcd is more suitable than pure lcd interface. lcd_framebuffer.c can work immediately with LVGL porting. If we give up lcd_framebuffer.c, the porting on top of /dev/lcdx will become more complex to get the advanced UI experience.

Have you tried it and actually observed this issue? I'm using LCD + LVGL without any issues and is quite efficient. As I mentioned, if you have an LCD, using putarea() with LCD will be the best way. The extra buffer added by lcd_framebuffer.c does not add anything to this: LVGL already is drawing to its own internal buffer.

I don't want to extend this discussion much further: we can leave the LCD + framebuffer driver be. But please note that it seems you are concerned with efficiency when your assumption about LCD framebuffer being faster on LVGL is incorrect.

Reply:
> So, what I expect is that:
> 
>     1. For resource constraint system, user enable CONFIG_LCD_DEV and LVGL open /dev/lcdx
> 
>     2. For advanced UI, user enable CONFIG_LCD_FRAMEBUFFER and LVGL open /dev/fbx

Please see my response, LCD FRAMEBUFFER does not do anything useful with LVGL. In fact, it only requires unneeded memory.

Reply:
Yes, LVGL use the internal buffer to do the drawing, but there are many GUI library directly talk to the underlying "frame buffer" directly. In this case, lcd_framebuffer is very useful. Also the patch isn't complete from the implementation point: why lcd_dev support putarea but lcd_framebuffer not? unless you remove lcd_framebuffer.c from the code base,

Reply:
We can leave the LCD framebuffer as is. IMHO I think it isn't the best way to solve the issue you mention from an architectural standpoint, but I'm OK with leaving it.
With respect to adding putarea to LCD framebuffer, I can do that but note that I will not be able to test it.

Reply:
Ok, we will provide the patch for lcd framebuffer.

Reply:
Ok, thanks!